pkg/adsc2/client.go

```diff
@@ -0,0 +1,620 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adsc2
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/resource/v3"
+	"go.uber.org/atomic"
+	"golang.org/x/net/context"
+	"google.golang.org/genproto/googleapis/rpc/status"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"k8s.io/utils/set"
+
+	"istio.io/istio/pilot/pkg/xds"
+	v3 "istio.io/istio/pilot/pkg/xds/v3"
+	"istio.io/istio/pkg/adsc"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/slices"
+	"istio.io/istio/pkg/util/sets"
+)
+
+type resourceKey struct {
+	Name    string
+	TypeURL string
+}
+
+func (k resourceKey) String() string {
+	return k.TypeURL + "/" + k.Name
+}
+
+func (k resourceKey) ShortName() string {
+	return v3.GetShortType(k.TypeURL) + "/" + k.Name
+}
+
+type keySet = sets.Set[resourceKey]
+
+type resourceNode struct {
+	// Parents of the resource. If nil, this is explicitly watched
+	Parents keySet
+	// Children of the resource
+	Children keySet
+}
+
+type HandlerContext interface {
+	RegisterDependency(typeURL string, resourceName ...string)
+	Reject(reason error)
+}
+
+var _ HandlerContext = &handlerContext{}
+
+// HandlerContext provides an event for a single delta response, allowing handlers to react to it.
+// Operations done in the handler may be batched together with other handler's.
+type handlerContext struct {
+	sub  keySet
+	nack error
+}
+
+func (h *handlerContext) RegisterDependency(typeURL string, resourceName ...string) {
+	if h.sub == nil {
+		h.sub = make(keySet)
+	}
+	for _, r := range resourceName {
+		key := resourceKey{
+			Name:    r,
+			TypeURL: typeURL,
+		}
+		h.sub.Insert(key)
+	}
+}
+
+func (h *handlerContext) Reject(reason error) {
+	h.nack = reason
+}
+
+// DeltaADSConfig for delta ADS connection.
+type DeltaADSConfig struct {
+	// TODO mcp, reconnect
+	*adsc.Config
+}
+
+type Client struct {
+	config   *DeltaADSConfig
+	handlers map[string]interface{}
+	// Map of resources key to its node
+	tree map[resourceKey]resourceNode
+
+	xdsClient discovery.AggregatedDiscoveryService_DeltaAggregatedResourcesClient
+	conn      *grpc.ClientConn
+
+	// initialWatches is the list of resources we are watching on startup
+	initialWatches []resourceKey
+
+	// sendNodeMeta is set to true if the connection is new - and we need to send node meta
+	sendNodeMeta atomic.Bool
+
+	mutex sync.RWMutex
+	// Last received message, by type
+	received        map[string]*discovery.DeltaDiscoveryResponse
+	deltaXDSUpdates chan *discovery.DeltaDiscoveryResponse
+
+	resourceCache sync.Map
```

Comment:
We store the value to manage the resource when it is removed; otherwise, from the DeltaDiscoveryResponse, we would only know the name of the removed resource. Currently, we watch specific types/resources instead of all resources to minimize memory usage. Would it be better to make the cache optional for cases where tracking deleted resources is not necessary?

Reply:
>otherwise, from the DeltaDiscoveryResponse, we would only know the name of the removed resource.

I am not sure isn't this satisfying

Reply:
I've adde an option, and also some explainations.

https://github.com/istio/istio/blob/1dd980cab5451814af3442e4fee6a413efd89636/pkg/adsc/delta.go#L96-L129

 Can this address your concern?